libmamba/src/api/configuration.cpp

```diff
@@ -1005,24 +1005,23 @@ namespace mamba
 
         void envs_dirs_hook(const Context& context, std::vector<fs::u8path>& dirs)
         {
-            // Check that "root_prefix/envs" is already in the dirs,
-            // and append if not - to match `conda`
-            fs::u8path default_env_dir = context.prefix_params.root_prefix / "envs";
-            if (std::find(dirs.begin(), dirs.end(), default_env_dir) == dirs.end())
-            {
-                dirs.push_back(default_env_dir);
-            }
-
-            // Also add all the directories in the environment variable `CONDA_ENVS_PATH`.
+            // Prepend all the directories in the environment variable `CONDA_ENVS_PATH`.
             auto conda_envs_path = util::get_env("CONDA_ENVS_PATH");
+
             if (conda_envs_path)
             {
+                if (util::get_env("CONDA_ENVS_DIRS"))
+                {
+                    LOG_ERROR << "CONDA_ENVS_DIRS and CONDA_ENVS_PATH both set. Must declare only one (prefer CONDA_ENVS_DIRS)";
+                    throw std::runtime_error("Aborting.");
+                }
+
                 auto paths_separator = util::pathsep();
 
                 auto paths = util::split(conda_envs_path.value(), paths_separator);
-                for (auto& p : paths)
-                {
-                    dirs.push_back(fs::u8path(p));
+                for (auto it = paths.rbegin(); it != paths.rend(); ++it)
+                {  // prepend conda_envs_path to dirs while maintaining order
+                    dirs.insert(dirs.begin(), fs::u8path(*it));
```

Comment:
This will come with `paths.size()` reallocation. Could an alternative be:

```c++
std::move(dirs.begin(), dirs.end(), std::back_inserter(paths));
dirs = paths;
```

?

Reply:
That gets a bit messy since one is a vector of u8path and one is a vector of string.  How about `reserve()` and `emplace()` instead?

Reply:
I guess you meant `resize`, in that case yes (I have seen your comment).

Reply:
My C++ is admittedly very rusty but I believe we do want `reserve` instead of `resize` (`resize` would default-construct elements at the end of the list).

Reply:
I meant, if one is emplacing at `dirs.begin()`, all the elements need to be moved toward the end so this is $\mathcal{O}\left(n^2\right)$, where $n$ is the number of elements.

I guess we could use a `std::dequeue` but we still need to convert it at the end anyway.

Reply:
Sure - the downside of a vector is the added expense for insertions.  Given that the function only gets executed once and the number of elements is probably going to be small (I would expect 0-2 in nearly all cases), I'd argue we should investigate further optimization (e.g. replacing it across the board with a `std::list` or `std::deque`) in a separate PR

Reply:
Yes, I also agree, also sometimes operations are generally faster with `std::vector` despite a worst complexity of operations thanks to cache/memory locality.

We can merge this PR once this the conflicts are resolved. :slightly_smiling_face: